src/qml/qmlwaveformrenderer.cpp

```diff
@@ -0,0 +1,243 @@
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererPreroll::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* renderer = new WaveformRendererPreroll(
+            waveformWidget, m_position);
+    renderer->setColor(m_color);
+    connect(this,
+            &QmlWaveformRendererPreroll::colorChanged,
+            renderer,
+            &WaveformRendererPreroll::setColor);
+    return QmlWaveformRendererFactory::Renderer{renderer, renderer};
+}
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererRGB::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* renderer = new WaveformRendererRGB(waveformWidget, m_position, m_options);
+
+    renderer->setAxesColor(m_axesColor);
+    renderer->setLowColor(m_lowColor);
+    renderer->setMidColor(m_midColor);
+    renderer->setHighColor(m_highColor);
+    connect(this,
+            &QmlWaveformRendererRGB::axesColorChanged,
+            renderer,
+            &WaveformRendererRGB::setAxesColor);
+    connect(this,
+            &QmlWaveformRendererRGB::lowColorChanged,
+            renderer,
+            &WaveformRendererRGB::setLowColor);
+    connect(this,
+            &QmlWaveformRendererRGB::midColorChanged,
+            renderer,
+            &WaveformRendererRGB::setMidColor);
+    connect(this,
+            &QmlWaveformRendererRGB::highColorChanged,
+            renderer,
+            &WaveformRendererRGB::setHighColor);
+
+    renderer->setAllChannelVisualGain(m_gainAll);
+    renderer->setLowVisualGain(m_gainLow);
+    renderer->setMidVisualGain(m_gainMid);
+    renderer->setHighVisualGain(m_gainHigh);
+    connect(this,
+            &QmlWaveformRendererRGB::gainAllChanged,
+            renderer,
+            &WaveformRendererRGB::setAllChannelVisualGain);
+    connect(this,
+            &QmlWaveformRendererRGB::gainLowChanged,
+            renderer,
+            &WaveformRendererRGB::setLowVisualGain);
+    connect(this,
+            &QmlWaveformRendererRGB::gainMidChanged,
+            renderer,
+            &WaveformRendererRGB::setMidVisualGain);
+    connect(this,
+            &QmlWaveformRendererRGB::gainHighChanged,
+            renderer,
+            &WaveformRendererRGB::setHighVisualGain);
+    return QmlWaveformRendererFactory::Renderer{renderer, renderer};
+}
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererBeat::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* renderer = new WaveformRenderBeat(
+            waveformWidget, m_position);
+    renderer->setColor(m_color);
+    connect(this, &QmlWaveformRendererBeat::colorChanged, renderer, &WaveformRenderBeat::setColor);
+    return QmlWaveformRendererFactory::Renderer{renderer, renderer};
+}
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererMarkRange::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* renderer = new WaveformRenderMarkRange(
+            waveformWidget);
+
+    for (auto* pMark : m_ranges) {
```

Comment:
Shouldn't this be const?

Reply:
use `std::as_const` too. Otherwise it may still detach the container.